ToolChain/Compiler/Compiler.py

```diff
@@ -0,0 +1,1877 @@
+        its own restrictions based on what input it expects next. This parse method is essentially the controller that
+        directs input.
+        :param char: char, Individual character read from input file
+        :param output: file, output file to write to
+        :return:
+        """
+
+        if self.state == 0:
+            self.state0(char, output)
+
+        elif self.state == 1:
+            self.state1(char, output)
+
+        elif self.state == 2:
+            self.state2(char, output)
+
+        elif self.state == 3:
+            self.state3(char, output)
+
+        elif self.state == 4:
+            self.state4(char, output)
+
+        elif self.state == 5:
+            self.state5(char, output)
+
+        elif self.state == 6:
+            self.state6(char, output)
+
+        elif self.state == 7:
+            self.state7(char, output)
+
+        elif self.state == 8:
+            self.state8(char, output)
+
+        elif self.state == 9:
+            self.state9(char, output)
+
+        elif self.state == 10:
+            self.state10(char, output)
+
+        elif self.state == 11:
+            self.state11(char, output)
+
+        elif self.state == 12:
+            self.state12(char, output)
+
+        elif self.state == 13:
+            self.state13(char, output)
+
+        elif self.state == 14:
+            self.state14(char, output)
+
+        elif self.state == 15:
+            self.state15(char, output)
+
+        elif self.state == 16:
+            self.state16(char, output)
+
+        elif self.state == 17:
+            self.state17(char, output)
+
+        elif self.state == 18:
+            self.state18(char, output)
+
+        elif self.state == 19:
+            self.state19(char, output)
+
+    def state0(self, char, output):
+        """
+        First step in parsing data. At this step, we begin to read the method header. We expect to read the return data
+        type.
+        :param char: char, Individual character read from input file
+        :param output: file, output file to write to
+        :return:
+        """
+
+        if char in IGNORE_CHARS and self.expectFlag == 0:
+            pass
+        elif char in IGNORE_CHARS and self.expectFlag == 1:
+            if self.currentType in ACCEPTED_TYPES:
+                self.state = 1
+                self.expectFlag = 0
+            else:
+                raise ValueError("Incorrect return type for method declaration at line {}.".format(self.lineno))
+        else:
+            self.currentType += char
+            self.expectFlag = 1
+
+    def state1(self, char, output):
+        """
+        Here we expect to read the method's name. Once we reach a space or an opening parentheses, we add the method
+        to the methodlist along with its data type.
+        TODO: figure out a way to have the main method printed first in the .casm file
```

Comment:
A tree structure starting from main would work ;)
It would also deal with the issue of dead code (in the case a full function is not called)

Reply:
Obviously something would also need to be worked out for libraries in that case.